### ISSUE TITLE
Trying to add manual SMS feature.

### ISSUE THREAD

**Issue by RamkrishnaMondal** [jekkos/opensourcepos#472]
I am using  manual sms sending tool written in php to send messages to my customers. The tool is working fine if i use from my localhost root directory ( ie var/www/ ). But showing lot of errors while trying to use it as a' module' from within the OSPOS, as the tool is written to use 'mysql' not 'mysqli' driver.

Can any one help converting the mysql functions to mysqli function? 

![screenshot from 2016-04-10 14 06 54](https://cloud.githubusercontent.com/assets/14151032/14408928/83675892-ff25-11e5-928b-8fb2c7b57ed3.png)

Using 'error_reporting(0); it looks like below : 

![screenshot from 2016-04-10 17 53 58](https://cloud.githubusercontent.com/assets/14151032/14410169/c786eafe-ff45-11e5-8d0a-bf46e201d7ff.png)

**Comment by jekkos** [comment]
This code looks like an ugly hack. You should try to use CI models + controllers, add new entries in modules table and all should be fine.

**Comment by jekkos** [comment]
Just read CI documentation a bit and you'll find out how it works

**Comment by RamkrishnaMondal** [comment]
Thanks to @jekkos, as per his advice using CI models I have done the following:

![screenshot from 2016-04-15 20 43 30](https://cloud.githubusercontent.com/assets/14151032/14570856/fb39b2ec-0363-11e6-88da-8a9813569a6d.png)

![screenshot from 2016-04-15 20 42 57](https://cloud.githubusercontent.com/assets/14151032/14570791/a4336dd0-0363-11e6-8629-8e418821b564.png)
![screenshot from 2016-04-15 20 42 33](https://cloud.githubusercontent.com/assets/14151032/14570794/aaa703fc-0363-11e6-8720-f6e6347c3982.png)
![screenshot from 2016-04-15 20 42 07](https://cloud.githubusercontent.com/assets/14151032/14570803/b16a0dce-0363-11e6-86f2-85546b558f03.png)
![screenshot from 2016-04-15 20 41 39](https://cloud.githubusercontent.com/assets/14151032/14570862/07b017fa-0364-11e6-89fe-74cbcb890d46.png)


Wanted to use the Modal dialogue like this one:

![screenshot from 2016-04-15 23 38 01](https://cloud.githubusercontent.com/assets/14151032/14570906/355e0284-0364-11e6-9d8a-f88aee9d835a.png)

But the form is not submitting from within the Modal dialogue ....... if @daN4cat  can help regarding this, as OSPOS uses edit/update forms within modal dialogue. 

Thanks.

**Comment by kdkrun** [comment]
Dear Ram, Your Interface is very beautiful... ! This is very user friendly rather then default version of OSPOS, I would like clone your system. Would you like to share it with us?

**Comment by daN4cat** [comment]
That's OSPOS 3 Jekkos and I are developing. It's in master nothing to clone.

**Comment by kdkrun** [comment]
What about SMS Module, Is it included in OSPOS 3

**Comment by daN4cat** [comment]
I'm afraid not, unless somebody contributes back.

**Comment by daN4cat** [comment]
@RamkrishnaMondal probably you need a controller function to take the input at submit time and send the SMS. Just check any other form and how the input is taken in and saved.

**Comment by RamkrishnaMondal** [comment]
@kdkrun  I shall share my work, when it gets completed, except the 'SMS-API' as I am using  closed source one. Hope You will be able to use any other API.

@daN4cat  As you can see the 'sms sender html form' is triggered from the people manage table. Then the Message controller fetches the data from db via model to populate the form fields, now the 'sender-form' posts the input data to the 'sms-api.php' and api connects with sms-gateway and sends the sms .

Problem is, if i use the form as 'plane html' or wrapped with 'jumbotron' class the sms is sending, if I use the form wrapped with 'modal-dlg' class ( like edit/update forms are using in ospos)  on clicking the 'submit' button, only the 'modal dialogue' is getting closed..... no sms is sending.

I think to submit the 'form data' correctly, from within the modal-dlg , we need customized Ajax / javascript function. But I don't know Ajax/js.

**Comment by RamkrishnaMondal** [comment]
Next to do is .... 
Send  sms automatically to the customer when payment is made at the time of registering sale, if a phone number exists.

**Comment by daN4cat** [comment]
@RamkrishnaMondal when you submit you should call the Messages controller (let's call it) send function. In that function you get the data from the form and call the 'sms-api.php'.

If you take an existing form like views\customers\form.php you see:

`<?php echo form_open('customers/save/'.$person_info->person_id, array('id'=>'customer_form', 'class'=>'form-horizontal')); ?>`

change the part after the open to invoke the send function from Messages controller, like this:

`<?php echo form_open('messages/send/', array('id'=>'item_form', 'enctype'=>'multipart/form-data', 'class'=>'form-horizontal')); ?>`

in your Messages controller you do:

function send()
{
	$message = array(
		'first_name'=>$this->input->post('first_name'),
		'last_name'=>$this->input->post('last_name'),
		'phone_number'=>$this->input->post('phone_number'),
		'message'=>$this->input->post('message')
	);

	// do something to send the message....
}

**Comment by daN4cat** [comment]
And to show something at the bottom to confirm everything was sent fine add at the end of function send:

```
if(<SMS send function result>)
{
	echo json_encode(array('success'=>true,'message'=>$this->lang->line('message_sent_successfully')));
}
else//failure
{	
	echo json_encode(array('success'=>false,'message'=>$this->lang->line('message_sent_failure')));
}
```

**Comment by daN4cat** [comment]
Of course you need to add the translations.

**Comment by RamkrishnaMondal** [comment]
Thanks @daN4cat for the help.  Form data is submitting now, from within the modal , without using 'send' function " <?php echo form_open('messages/' " does the trick.

I wish to use the 'OSPOS' inbuilt notification like as other modules do.... will have to find out a way to invoke it. If you can, please help regarding this. I also need suggestion regarding editing Sales controller  for enabling the automation.

**Comment by daN4cat** [comment]
Please add the send function, don't use the default path. Do as I suggested.

**Comment by daN4cat** [comment]
For the feedback message check other forms and views. Also make sure you have the footer included in your view.

**Comment by RamkrishnaMondal** [comment]
Tried to use the 'send' function but after sending the sms it brings me to a blank page.

Footer is already enabled to the default view file of the "messages' module but after sending sms, when it redicts me  to the messages module default view page..... footer is not getting loaded, header loads though.

**Comment by daN4cat** [comment]
You need to reload your view in your send. See other controllers examples.

**Comment by daN4cat** [comment]
Check _reload function in Sales controller for instance.

**Comment by kdkrun** [comment]
My Dear Friends, If you need a SMS API with sender ID (Number Masking Feature with 11 Character String) to test sending SMS, I can provide it for free with some SMS Credits. Please let me know. I like to help you someway I can.

**Comment by daN4cat** [comment]
For the feedback you need js in your view to handle the json message.

**Comment by daN4cat** [comment]
Any progress with this?

**Comment by RamkrishnaMondal** [comment]
@jekkos    Mannual sending is working fine now, automation is still not implemented. If you wish to take care of the automation part, I can upload my works here as a zip folder. Otherwise I need some time to complete the automation.

**Comment by daN4cat** [comment]
Work in two steps, 1) manual feature and we merge  then 2) you can work on the automated

It would be better if you have it in your repository on GitHub and you go for a Pull Request. It leaves even your name in the history attached to this feature, which is a nice thing.

**Comment by RamkrishnaMondal** [comment]
Sorry for the typo ...it was for daN4cat....corrected.

**Comment by RamkrishnaMondal** [comment]
I don't know how to create a github repo on my local machine (ubuntu14.0.4.4)....

**Comment by daN4cat** [comment]
@RamkrishnaMondal well, weekends are there to spend time to do something fun right? :-)

This week end topic for you is: learn GitHub. You can only gain from this because it's widely used and it will look good on your CV.
Since you managed to add a feature to OSPOS, you have all the skills to learn GitHub.

Make sure you have a fork of jekkos/opensourcepos first, so you can work from your repository first.

To fork the repo just go here: https://github.com/jekkos/opensourcepos and then click the fork icon on the top right.
That will create a copy of latest master under your name: https://github.com/RamkrishnaMondal/opensourcepos

Then see something like this: http://www.ubuntumanual.org/posts/393/how-to-setup-and-use-github-in-ubuntu or search on YouTube for some tutorial on GitHub that explains step by step how to on Ubuntu in terms of getting code from your repo to your local machine and change code and commit to your repo.

Make sure you have a backup of the changes you made. I know you might get crazy in merging what you have done with the latest master, but no pain no gain :-)

**Comment by RamkrishnaMondal** [comment]
Pull request sent....... hope, I have not messed up things.

**Comment by daN4cat** [comment]
Good for the PR, but I'm afraid the implementation is not as we discussed.
It has to follow the general pattern of OSPOS, so sending of message needs to happen via the controller calling the sms-api, not from the view.

I cannot merge this code.

**Comment by RamkrishnaMondal** [comment]
Well, i shall try to modify as per your suggestion.

**Comment by RamkrishnaMondal** [comment]
BTW,  where should i place the api file ?

**Comment by ekardian** [comment]
more brains for test, share a link for comunity test :)

**Comment by RamkrishnaMondal** [comment]
It is already in my repo and working just fine for me. Only problem is, it is not as per OSPOS coding pattern.

**Comment by odiea** [comment]
In the database you need to change ('quantity_decimals', '0'); to ('quantity_decimals', '0'),
('quantity_decimals', '0');
('msg_msg',''),
('msg_uid', 'Put api userName'),
('msg_pwd', 'Put api password');

**Comment by RamkrishnaMondal** [comment]
Done... thanks for pointing out.

**Comment by daN4cat** [comment]
@RamkrishnaMondal if the api file you talk about is the service you use and it's their code, put it in third_party.

It the api file you talk about is the one you call and needs to be implemented to interface the service you use, it's better to place it under helpers calling it sms_helper.php.
To make the helper available you need to go to application/config/autoload.php and in $autoload['helper'] = array(...); add 'sms' at the end.

**Comment by RamkrishnaMondal** [comment]
@daN4cat  I have changed the controller and put the api as a lib file. 
```
<?php  if ( ! defined('BASEPATH')) exit('No direct script access allowed');

require_once ("Secure_area.php");
class Messages extends Secure_area
{
	function __construct()
	{
		parent::__construct('messages');
	}
	
	public function index()
	{
		$data['controller_name'] = $this->get_controller_name();
		$this->load->library('sms_api_lib');
		$this->load->view('messages/sms');
	}
	/*
	Loads the sms sender form
	*/
	function view($person_id=-1)
	{ 
		$data['person_info']=$this->Person->get_info($person_id);
		$this->load->view('messages/sms-sender',$data);
	}

	function send()
	{

		$params = array(
		'uid'=>$this->input->post('uid'),
		'pwd'=>$this->input->post('pwd'),
		'phone'=>$this->input->post('phone'),
		'msg'=>$this->input->post('msg')
	          );
                
		$this->load->library('sms_api_lib', $params);
		$this->sms_api_lib->send_sms('uid','pwd','phone','msg'); //'send_sms' is a inbult function of sms-api//
		$this->load->view('messages/sms');
	}	

}
?>``
```

**Comment by RamkrishnaMondal** [comment]
@daN4cat How do I fetch your recent changes to upstream master to my repo ?

**Comment by RamkrishnaMondal** [comment]
Searching for a  SMS-API with some trial credit which works Globally. The Api I am using is only for Indian  Mobile numbers.

**Comment by RamkrishnaMondal** [comment]
SMS integration is ready now, as per ospos coding pattern. Meanwhile  @daN4cat  has committed some changes upstream ..... i shall have to rearrange my codes and re-testing again.

**Comment by daN4cat** [comment]
Yes, that will always happen as everybody progresses different topics in parallel.

However I would stop now waiting for different things to come in or consider working on the Chart part on a branch, we'll see.

**Comment by RamkrishnaMondal** [comment]
I will push my work today so that you and others can test, If everything seems ok , I will send a PR.

**Comment by daN4cat** [comment]
Sounds good, thanks.

**Comment by RamkrishnaMondal** [comment]
SMS feature pushed to my fork repo,  please download and try a fresh install and report.

Note: Although when you click Submit button "Message submitted successfully " appears, you will need to replace the libraries/Sms.php file with your working sms-api file to 'actually' send the message.

**Comment by daN4cat** [comment]
Thanks, I'll have a look into it this evening.

**Comment by kdkrun** [comment]
@RamkrishnaMondal 
Here are some API that you can use to test, I have assigned 22 credits to this account, please let me know if you want more.

Please find the attached word document file for API Details
Login URL: http://204.93.168.114/sms/
User Name: Samadie-Tec
Password: welcome
Sender ID: Samadie-Tec
Incomming No: 94115945046

Please find an below API details

Sending API :

http://204.93.168.114/send_sms?username=kdktec&password=welcome&src=94115931566&dst=94702042691&msg= Client%20Hi This Is Test SMS&dr=1


API for delivery report.:

http://204.93.168.114/send_sms?username=kdktec&password=welcome&fdt=1/1/2015&tdt=20/1/2015&dr=2


Credit Balance :

http://204.93.168.114/send_sms?username=kdktec&password=welcome&dr=3


Numberwise Delivery Report Api

http://204.93.168.114/send_sms?username=kdktec&password=welcome&src=94115931566&dst=94702042691&dr=4

Please change highlight area

Please find the attached word document file for API Details
[New API Details.docx](https://github.com/jekkos/opensourcepos/files/236797/New.API.Details.docx)

**Comment by RamkrishnaMondal** [comment]
@kdkrun 
I can see there are 2 mobile nos in your api (src, dst),. Which one for recipient and which one for sender?

**Comment by RamkrishnaMondal** [comment]
@kdkrun 
Please check ...... can not send sms, even from http://204.93.168.114/sms/

**Comment by kdkrun** [comment]
src mean source numbers it is default telephone number I have assign to this sender ID called "Samadi-Tec" you no need to change this src number
dst is destination number you can assign php variable to this, probably customer telephone number

I have check this in my country, I mean for Sri Lanka and its works fine. I didn't check it for global telephone numbers

**Comment by RamkrishnaMondal** [comment]
I have tried to send  a sms from your website to your number ie 94115945046. It shows error.

![screenshot from 2016-04-27 19 10 34](https://cloud.githubusercontent.com/assets/14151032/14854444/60983d5a-0cad-11e6-8486-79317ce9fa8e.png)

**Comment by kdkrun** [comment]
Yes, You're right. I have check this issue with my service provider. They have told me to send an email and they will fix this issue before next Thursday. Until then please use following account information to test your system.
User Name : SisiraLearn
Password : welcome
Sender Id : SisiraLearn
Incomming No: 94115931553

Please use prevoius API configuration with this new user name, password and incoming no

**Comment by kdkrun** [comment]
@RamkrishnaMondal 


Yes, You're right. I have check this issue with my service provider. They have told me to send an email and they will fix this issue before next Thursday. Until then please use following account information to test your system.

User Name : SisiraLearn
Password : welcome
Sender Id : SisiraLearn
Incomming No: 94115931553

Please use prevoius API configuration with this new user name, password and incoming no
I have change previous account password so please use this account to test, I have assigned 21 SMS credits. If you need more, Please let me know.
Best regards! Ram

**Comment by daN4cat** [comment]
OK, I rework quite some parts of the original work to make sure the feature fits the rest of OSPOS properly.

First of all you must run the 2.4 to 3.0 upgrade script in order to add the Messages module to OSPOS otherwise there is no way it can be visible.

Secondly, without implementing an API to the service provider in application/library/sms.php the feature cannot work. See the couple of example in that code to see how things could work.

After having done that, the permission Messages need to be added to the admin user first in order to allow them to config the feature under Store config, Message tab.

Once that's done, go to Message view and type a phone number and a message and see if it works.

If all fine, going to Customer, Supplier, Employee and clicking the Mobile phone icon on any person will open a window to send a text to that persion.

At the moment there is no support of multiple selections and text sending. But should be doable in a similar manner as bulk edit. Maybe later that can be added.

**Comment by daN4cat** [comment]
And thank you to @RamkrishnaMondal to have initiated this work.

There is more to do if they want to have automatic alerts and etc. But I leave it to him to progress.
I might need to retouch few more bits later.

**Comment by odiea** [comment]
Thanks to both of you for all the work on this new feature. I am sure many will appreciate it.

**Comment by RamkrishnaMondal** [comment]
Thanks to @jekkos  for insisting me to learn CI basics to understand the OSPOS coding pattern.

Thanks to @daN4cat for constantly following up my work, suggesting , helping on various points, cleaning up codes and rewriting various parts. Without his constant support, this feature addition would not have been possible.  

Also thanks to @kdkrun for providing the api code and sms credits.

**Comment by RamkrishnaMondal** [comment]
@kdkrun   I think you should talk to your service provider, you need more info from them to get connected to their gateway. 
There is no universal rule to connect to the sms gateways, every service provider uses their own rule to by pass CAPTCHA and other security filters, that's the problem.

My suggestion is,  send them the sms lib of ospos and request them to modify the code as per their connection rules and use that modified lib.

**Comment by daN4cat** [comment]
Would be nice to have a generic way to interface those providers, but I'm afraid it is not.
Maybe we could have a list of common ones and a dropdown selection from Store Config so we avoid having people to ask how to add common ones.

That could be another addition, in the examples I have added a UK one. I could check what are the cheapest ones and add the php part in the file and maybe have this dropdown selection.
I need to think about what would be the best way to achieve that.

**Comment by RamkrishnaMondal** [comment]
Provided that, their api code is open sourced, we should not include any proprietary code with this opensource software.

**Comment by daN4cat** [comment]
Added a translation that somehow was left out and added glyphicons.

**Comment by daN4cat** [comment]
I would close this if no objections.

**Comment by daN4cat** [comment]
I close as there is already #211 for any more advanced feature.

**Comment by jekkos** [comment]
I had a quick look at this new module and it looks great overall. One remark might be the responsiveness of the screen. The width of the jumbotron container is set to 60% for all screen sizes which might not be approrpriate once you look at it on a smartphone / tablet.. 

We might want to define some kind of guideline for this type of screen (like break this rule once you go below a certain width). Same thing eg for config section.

**Comment by daN4cat** [comment]
I checked already the responsiveness at the time I touched the code and it was fine resizing the window. I countercheched on my mobile and the page is fine.
I don't perceive this as an issue at the moment.

**Comment by kdkrun** [comment]
Dear Friend, I am trying to add SMS API setting to OSPOS 3.0, Where we should changed $originator values. Following is my SMS API Url in my sms.php
$url = "http://204.93.168.114/SMS/send_sms?username=$username&password=$password&src=94115931566&dst=$phone&msg=$message&dr=1";

Can some one help me to put following values to above url?
http://204.93.168.114
username=kdktec
password=welcome
SMS Sending Number=94115931566

I hope that user name and password are put in to Setting section in OSPOS, But where should I put SMS Server IP (http://204.93.168.114) and Sending Number (94115931566)

Please help me in this regards.
Thank you.

**Comment by daN4cat** [comment]
You need to configure the values in Store Config / Messages.
The $originator is the Sender ID, use your phone number there (maybe we should change sender id to be originator?).

IP is not configurable because each service provider has a different syntax for the URL and how to send the message, so that cannot be generic.

For your case edit application/libraries/Sms.php to be:

`
		else
		{
			$response = TRUE;

			$url = "https://204.93.168.114/send_sms?username=$username&password=$password&src=$originator&dst=$phone&msg=$message&dr=1";
			 
			$c = curl_init(); 
			curl_setopt($c, CURLOPT_RETURNTRANSFER, 1); 
			curl_setopt($c, CURLOPT_URL, $url); 
			$response = curl_exec($c); 
			curl_close($c);
		}
`

**Comment by kdkrun** [comment]
Thank you friend, I have config as you say, but system always say that message successfully send, but its not. Herewith I'm attaching my sms.php file. can anyone check this for me.
I also attaching my SMS API setting with actual user name and passwords. and my pos system link is http://pos.kdktec.net
User Name: admin
Password: pointofsale
I have checked that following API SMS sending URL is working fine and I received the message.
http://204.93.168.114/SMS/send_sms?username=kdktec&password=kdkrun@3665555&src=94115931566&dst=94773204715&msg=Client%20Hi This Is Test SMS&dr=1
[SMS API Settings for KDKTEC.txt](https://github.com/jekkos/opensourcepos/files/281946/SMS.API.Settings.for.KDKTEC.txt)
[files.zip](https://github.com/jekkos/opensourcepos/files/281948/files.zip)

**Comment by daN4cat** [comment]
@kdkrun ok, I believe I know what the problem is.

Before $url = ...  line, add this `$message = rawurlencode($message);`

The message needs to have %20 and not white spaces and similar stuff.

Sorry if I flooded somebody with messages but I needed to test few times to make sure it was fine.

**Comment by daN4cat** [comment]
Committed a code change addint the rawurlencode of the message before being added to url and sent.

**Comment by kdkrun** [comment]
My Dear Friend Thank you for your wonderful help in this regards. Its works... I can't thank you enough... If you anyone need SMS API for develop this OS POS I would like to give it for free... please let me know.... Thank you again...

**Comment by daN4cat** [comment]
Glad to hear that works now.

**Comment by jonathanprieto99** [comment]
Hi,
I have the last versión of OSPOS installed on a LAMP with ubuntu 16.04.
I am looking for a tutorial on how to configure the sms API, which service to use and how to install it.
Could someone help me please. 
Thank you

**Comment by daN4cat** [comment]
I'm afraid there is no tutorial. Just go to application/libraries/Sms_lib.php and add your SMS gateway code there. Also search close issues as it was discussed few times.

**Comment by jonathanprieto99** [comment]
But could you say me how please ?

**Comment by daN4cat** [comment]
Have you opened that file and read the content?

**Comment by hasansemi** [comment]
How to setup this code in sms_lib.php

$curl = curl_init();

curl_setopt_array($curl, array(
  CURLOPT_URL => "http://api.msg91.com/api/v2/sendsms",
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => "",
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 30,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => "POST",
  CURLOPT_POSTFIELDS => "{ \"sender\": \"SOCKET\", \"route\": \"4\", \"country\": \"91\", \"sms\": [ { \"message\": \"Message1\", \"to\": [ \"98260XXXXX\", \"98261XXXXX\" ] }, { \"message\": \"Message2\", \"to\": [ \"98260XXXXX\", \"98261XXXXX\" ] } ] }",
  CURLOPT_SSL_VERIFYHOST => 0,
  CURLOPT_SSL_VERIFYPEER => 0,
  CURLOPT_HTTPHEADER => array(
    "authkey: ",
    "content-type: application/json"
  ),
));

$response = curl_exec($curl);
$err = curl_error($curl);

curl_close($curl);

if ($err) {
  echo "cURL Error #:" . $err;
} else {
  echo $response;
}

**Comment by hasansemi** [comment]
The major confusion is 
CURLOPT_POSTFIELDS => "{ "sender": "SOCKET", "route": "4", "country": "91", "sms": [ { "message": "Message1", "to": [ "98260XXXXX", "98261XXXXX" ] }, { "message": "Message2", "to": [ "98260XXXXX", "98261XXXXX" ] } ] }",

how to replace it .
the above share code work properly with my php script

**Comment by hasansemi** [comment]
@daN4cat @jekkos

**Comment by RamkrishnaMondal** [comment]
Look here: https://api.msg91.com/apidoc/samplecode/php-sample-code-send-sms.php
